### ISSUE TITLE
Driver-side debugging features

### ISSUE THREAD

**Issue by Jasper-Bekkers** [KhronosGroup/Vulkan-Docs#1930]
@Tobski requested I post this here

So day to day we encounter a lot of these kind of crashes during development, driver doesn't really matter and neither does vendor, it's a common pattern among all. Typically this is because one of the drivers doesn't recognize an opcode or sequence of opcodes.
Usually the only way to debug this, would be to strip the shader down, bisect stuff and hope you find the sequence and then write something else that works, file a bug, or both.

Obviously all software has bugs, but the way to debug this kind of problem is... Barbaric. We discuss a lot a out the advantages of having a debugger over printf debugging, but this is stochastic debugging if you ask me. Having symbols here would already be a massive step forward (usually if you have them you can correlate the callstack with the AST for example).

![image](https://user-images.githubusercontent.com/49594/188130161-b2fb9b93-2c62-4f85-b461-623a52703975.png)

In this case it's a crash into an Nvidia driver, but a typical case when iterating on a shader is that we just get a access violation somewhere directly inside the driver, and with no real way of debugging this from our point of view, we just end up manually permuting the shader that we're working on.

Any kind of additional information that the driver can output here would be useful; the current OpCode being processed, a detailed log that can be enabled (through an env-var, or otherwise) that printf's a trace of what's going on, some asserts and the text they include etc etc would all be extremely helpful here.

My expectation is that most drivers have these kind of tools build in already to help driver engineers assist in debugging these kinds of issues already, mostly as a normal part of their development processes. For example; typical clang based tooling could turn on and display it's asserts.

**Comment by Tobski** [comment]
This seems spiritually similar to https://registry.khronos.org/OpenGL/extensions/KHR/KHR_debug.txt, though most of that extension's functionality got put into VK_EXT_debug_utils. Adding a narrower "tell me what you already know" debugging extension for drivers (and layers?) definitely seems like it could be useful, and for the specific point of pipeline creation we can probably give you a bit more direct information (e.g. "this pipeline failed with this debug logging info", possibly even with an OpCode, at least in some cases).

**Comment by pdaniell-nv** [comment]
There was a time in the Vulkan 1.0 days our driver would output helpful messages to the debug callbacks and return VK_ERROR_INVALID_SHADER_NV, but with the rise of SPIR-V VVL, and our behavior being non standard, we stopped supporting it. Maybe we should revisit this behavior formally in the Vulkan standard. It might be a huge amount of work for implementations that have been built assuming SPIR-V is always valid so have zero infrastructure for error handling and reporting.

**Comment by ShabbyX** [comment]
> > There was a time in the Vulkan 1.0 days our driver would output helpful messages to the debug callbacks and return VK_ERROR_INVALID_SHADER_NV, but with the rise of SPIR-V VVL, and our behavior being non standard, we stopped supporting it. Maybe we should revisit this behavior formally in the Vulkan standard. It might be a huge amount of work for implementations that have been built assuming SPIR-V is always valid so have zero infrastructure for error handling and reporting.
> 
> This is for cases where the SPIRV is valid (spirv-val takes them and Dxc has output them) but the driver has a bug. We find these cases to be quite common but difficult to debug.

How do you expect the driver to tell you it's about to crash due to its own bug when handling some opcode?

**Comment by Jasper-Bekkers** [comment]
> > > There was a time in the Vulkan 1.0 days our driver would output helpful messages to the debug callbacks and return VK_ERROR_INVALID_SHADER_NV, but with the rise of SPIR-V VVL, and our behavior being non standard, we stopped supporting it. Maybe we should revisit this behavior formally in the Vulkan standard. It might be a huge amount of work for implementations that have been built assuming SPIR-V is always valid so have zero infrastructure for error handling and reporting.
> > 
> > This is for cases where the SPIRV is valid (spirv-val takes them and Dxc has output them) but the driver has a bug. We find these cases to be quite common but difficult to debug.
> 
> How do you expect the driver to tell you it's about to crash due to its own bug when handling some opcode?

I expect it to at least be able to log out a trace on its way there, and preferably have asserts enabled that I can check or investigate. Most drivers we've worked with in the past were capable of this for IHVs internal development use cases, but they're also extremely useful for application developers.

**Comment by krOoze** [comment]
Should probably not happen much in the first place, otherwise it points to a lack of discipline with maintaining the Vulkan CTS.

**Comment by Jasper-Bekkers** [comment]
> These class of problems should probably not happen much in the first place, otherwise it points to a lack of discipline with maintaining the Vulkan CTS, which should test things preemptively and not when discovered by an end-user in a publically released driver.

Like I said we encounter these things on a daily basis, no need to doubt that.

Yes these are probably holes in the CTS tests or in the inhouse testing that hardware vendors do. To help cover those holes and fix issues, I'm asking for more debugging tools to become available ☺️

**Comment by krOoze** [comment]
Drivers may express problems with `VK_ERROR_VALIDATION_FAILED_EXT` or `VK_ERROR_UNKNOWN`, and I think they are allowed to write something to messengers if they wish so.

Not entirely convinced they should be mandated to do so generally. It would be treating the symptom rather than the disease.

**Comment by ShabbyX** [comment]
> Most drivers we've worked with in the past were capable of this for IHVs internal development use cases, but they're also extremely useful for application developers.

That's the key difference IMO here. Drivers most definitely have assertions, logs, and various other debugging features they'd use for development. I don't think we could realistically expect them to _ship_ with them.

I'm totally not opposed to asking vendors to provide debug builds of their drivers though, that would most certainly be helpful. It _is_ indeed very helpful when we see a crash in mesa for instance. Just the fact that we can symbolize the stack traces in fact is sufficient most of the times.

**Comment by Jasper-Bekkers** [comment]
> Drivers most definitely have assertions, logs, and various other debugging features they'd use for development. I don't think we could realistically expect them to _ship_ with them.

Why is that? The performance implications are solvable by returning different entry points or an entirely different dll. And if there's one thing I've learned over the past years, is that it's better to ask and let the vendors decided for themselves, rather then speculate on their behalf.

**Comment by Tobski** [comment]
@krOoze 

> Not entirely convinced they should be mandated to do so generally. It would be treating the symptom rather than the disease.

Most of these issues are going to be in weird corners that we can't catch easily with CTS, we are going to continue increasing coverage, but in the meantime, developers have a problem which we can solve relatively easily with something that isn't going to take months to put together. Ideally CTS would be perfect, sure, but that's rather an impossible goal.

@ShabbyX 

> Drivers most definitely have assertions, logs, and various other debugging features they'd use for development. I don't think we could realistically expect them to ship with them.

Two vendors have already said they are willing to do this, it certainly might be more costly on some platforms, but for debugging on a desktop machine this is unlikely to be a considerable cost.

**Comment by ShabbyX** [comment]
>  The performance implications are solvable by returning different entry points or an entirely different dll.

Yes, a debug build of the driver available for install would certainly be nice.

I'm not against any of this at all, in case that's not clear. If it were up to me, all drivers would be open source and we wouldn't be having this conversation either.

It certainly is helpful to ANGLE users that they see what it does, find bugs in it, or understand why a certain pattern leads to inefficiencies. So I understand perfectly well where you're coming from.

**Comment by krOoze** [comment]
I don't like the idea that with this we are basically subscribing to be unpaid testers and debuggers of proprietary drivers. If they are really getting these bugs on a daily basis, then this is fixing a proxy problem and helps avoid fixing the underlying problem.

**Comment by nanokatze** [comment]
Debug (debugoptimized?) builds of drivers would be nice to have, even if their (debug-only) diagnostics are output to just stderr rather than debug message handler in VK_EXT_debug_utils.

As far as I understand, drivers already are allowed and some already do print all kinds of diagnostics through VK_EXT_debug_utils message handler, e.g. all Mesa drivers do it through __vk_log and __vk_errorf, so no new functionality seems to be necessary?
